### PR TITLE
chore(notifications): enable strict mode

### PR DESCRIPTION
**Pull request by Samaritan1011001**
#### Description of changes

Enables strict mode for the Notifications packages along with the following changes. 

#### Description of how you validated changes
yarn test

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [ ] Tests are [changed or added](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#steps-towards-contributions)
- [ ] Relevant documentation is changed or added (and PR referenced)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/notifications/tsconfig.json [hunk 1]
```diff
@@ -15,7 +15,8 @@
 		"types": ["jest", "node"],
 		"esModuleInterop": true,
 		"resolveJsonModule": true,
-		"jsx": "react-native"
+		"jsx": "react-native",
+		"strict": true
```
Comment (AllanZhengYP) [inline packages/notifications/tsconfig.json:19]: You should also set `noImplicitAny` to true.

File: packages/notifications/tsconfig.build.json [hunk 2]
```diff
@@ -1,5 +1,9 @@
 {
```
Comment (cshfang) [inline packages/notifications/tsconfig.build.json:1]: Is this file still needed?

**Review by cshfang — DISMISSED** [review]

File: packages/notifications/tsconfig.build.json (hunk 2 above)
Comment (jimblanc) [reply packages/notifications/tsconfig.build.json:1]: Ditto, if you update `ts-coverage` to use `tsconfig.json` should be able to remove this

**Review by jimblanc — DISMISSED** [review]

File: yarn.lock [hunk 3]
```diff
@@ -13465,9 +13608,9 @@ react-native@0.72.3:
     yargs "^17.6.2"
 
 react-native@>=0.70:
-  version "0.72.5"
-  resolved "https://registry.yarnpkg.com/react-native/-/react-native-0.72.5.tgz#2c343fa6f3ead362cf07376634a33a4078864357"
-  integrity sha512-oIewslu5DBwOmo7x5rdzZlZXCqDIna0R4dUwVpfmVteORYLr4yaZo5wQnMeR+H7x54GaMhmgeqp0ZpULtulJFg==
+  version "0.72.6"
```
Comment (AllanZhengYP) [inline yarn.lock:13611]: Nit: Why do we need to upgrade RN?

File: packages/notifications/src/inAppMessaging/providers/pinpoint/utils/messageProcessingHelpers.ts [hunk 4]
```diff
@@ -114,17 +118,29 @@ async function isBelowCap({
...
+				sessionCount: getSessionCount(messageId),
+				dailyCount: await getDailyCount(),
+				totalCount: await getTotalCount(messageId),
+			};
 	} catch (err) {
 		logger.error('Failed to get message counts from storage', err);
+
+		// If there are no cached counts or there is an error,
+		// we default to 0 allowing all the messages to be eligible
+		return messageCounts;
```
Comment (jimblanc) [inline packages/notifications/src/inAppMessaging/providers/pinpoint/utils/messageProcessingHelpers.ts:141]: Nit: I think you can just return messageCounts in the try block above, remove this, and have the default return at the end of the function in the event of the error.

File: packages/notifications/src/inAppMessaging/providers/pinpoint/utils/messageProcessingHelpers.ts [hunk 5]
```diff
@@ -88,9 +88,13 @@ function normalizeMessages(messages: PinpointInAppMessage[]): InAppMessage[] {
 	return messages.map(message => {
 		const { CampaignId, InAppMessage } = message;
 		return {
-			id: CampaignId,
+			// Default to empty string in rare cases we don't have a campaignId
+			id: CampaignId ?? '',
 			content: extractContent(message),
-			layout: interpretLayout(InAppMessage.Layout),
+			// Default to TOP_BANNER layout in rare cases we don't have a Layout
```
Comment (jimblanc) [inline packages/notifications/src/inAppMessaging/providers/pinpoint/utils/messageProcessingHelpers.ts:94]: As a sanity check, is this a behavior change?

File: packages/notifications/src/inAppMessaging/providers/pinpoint/utils/helpers.ts [hunk 6]
```diff
@@ -218,9 +223,9 @@ export const clearMemo = () => {
 // - 1. the usage of MOBILE_FEED and OVERLAYS as values for message layouts are not leaked
 //      outside the Pinpoint provider
 // - 2. Amplify correctly handles the legacy layout values from Pinpoint after they are updated
-export const interpretLayout = (
-	layout: PinpointInAppMessage['InAppMessage']['Layout']
-): InAppMessageLayout => {
+
+// TODO(V6): Add exact type as PinpointInAppMessage['InAppMessage']['Layout']
```
Comment (jimblanc) [inline packages/notifications/src/inAppMessaging/providers/pinpoint/utils/helpers.ts:227]: Does this need to happen before GA?

File: packages/notifications/src/inAppMessaging/providers/pinpoint/utils/helpers.ts [hunk 7]
```diff
@@ -258,21 +263,23 @@ export const extractContent = ({
 			}
 			if (HeaderConfig) {
 				extractedContent.header = {
-					content: HeaderConfig.Header,
+					// Default to empty string in rare cases we don't have a Header value
```
Comment (jimblanc) [inline packages/notifications/src/inAppMessaging/providers/pinpoint/utils/helpers.ts:266]: Sanity check: Is this behavior change acceptable if returning undefined before?

File: packages/notifications/src/common/providers/pinpoint/types/config.ts [hunk 8]
```diff
@@ -0,0 +1,7 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+export interface PinpointProviderConfig {
```
Comment (jimblanc) [inline packages/notifications/src/common/providers/pinpoint/types/config.ts:4]: Is this type used anywhere (can't find in PR)? Also I think we already have this in core?

File: packages/notifications/package.json [hunk 9]
```diff
@@ -105,6 +103,7 @@
 		"@babel/plugin-proposal-class-properties": "^7.0.0",
 		"@babel/plugin-proposal-private-methods": "^7.0.0",
 		"@babel/plugin-proposal-private-property-in-object": "^7.0.0",
-		"metro-react-native-babel-preset": "^0.66.2"
+		"metro-react-native-babel-preset": "^0.66.2",
+		"@types/react-native": "0.70.0"
```
Comment (jimblanc) [inline packages/notifications/package.json:107]: I'll defer to Chris, but shouldn't the RN dev dep only be on the RN package?

Comment (cshfang) [reply packages/notifications/package.json:107]: Should be the case going forward but this may be tied up with testing the *.native files here

File: yarn.lock (hunk 3 above)
Comment (Samaritan1011001) [reply yarn.lock:13611]: running `yarn` updated this lock file so I didn't upgrade the version myself in any of the packages. Confused why it has a new version when it's pinned to 0.70 everywhere.

File: packages/notifications/src/inAppMessaging/providers/pinpoint/utils/messageProcessingHelpers.ts (hunk 4 above)
Comment (Samaritan1011001) [reply packages/notifications/src/inAppMessaging/providers/pinpoint/utils/messageProcessingHelpers.ts:141]: Thanks for the feedback. 
It needs to return the default 0s in two cases, when there is no messageId or when there is an error. 
Which is why I have a default value set in the beginning and return either actual value or default 0s depending on the situation. 
I did move the return statement outside the block to inside the try block but doesn't make much of a difference.

File: packages/notifications/src/inAppMessaging/providers/pinpoint/utils/messageProcessingHelpers.ts (hunk 5 above)
Comment (Samaritan1011001) [reply packages/notifications/src/inAppMessaging/providers/pinpoint/utils/messageProcessingHelpers.ts:94]: I think there would be a type error or such thrown previously if this was undefined but now we have a default value. So in a way yeah, it's a behavior change.

File: packages/notifications/src/inAppMessaging/providers/pinpoint/utils/helpers.ts (hunk 6 above)
Comment (Samaritan1011001) [reply packages/notifications/src/inAppMessaging/providers/pinpoint/utils/helpers.ts:227]: Thank you, updated.

File: packages/notifications/src/inAppMessaging/providers/pinpoint/utils/helpers.ts (hunk 7 above)
Comment (Samaritan1011001) [reply packages/notifications/src/inAppMessaging/providers/pinpoint/utils/helpers.ts:266]: Same as above. 
Additional note: this shouldn't really happen but in rare cases it does, we now have a default value as opposed to throwing an error due to undefined.
Open to suggestion on how we want to handle the absence of these values.

File: packages/notifications/src/common/providers/pinpoint/types/config.ts (hunk 8 above)
Comment (Samaritan1011001) [reply packages/notifications/src/common/providers/pinpoint/types/config.ts:4]: thanks for catching, removed.

File: yarn.lock (hunk 3 above)
Comment (Samaritan1011001) [reply yarn.lock]: reverted

**Review by jimblanc — APPROVED** [review]

**Review by AllanZhengYP — APPROVED** [review]
Looks good to me!✅

File: packages/notifications/__tests__/inAppMessaging/providers/pinpoint/apis/setConflictHandler.test.ts [hunk 10]
```diff
@@ -33,8 +34,8 @@ describe('setConflictHandler', () => {
 		mockNotifyEventListeners.mockClear();
 	});
 	it('can register a custom conflict handler', async () => {
-		const customConflictHandler = messages =>
-			messages.find(message => message.id === 'custom-handled');
+		const customConflictHandler = (messages: InAppMessage[]) =>
+			messages.find(message => message.id === 'custom-handled') ?? messages[0];
```
Comment (cshfang) [inline packages/notifications/__tests__/inAppMessaging/providers/pinpoint/apis/setConflictHandler.test.ts:38]: Nit: Possible to use `!` to assert non-null on the find? Just an outside concern that this masks some functionality change where `messages[0]` is being returned inadvertently but the test is still passing

File: packages/notifications/__tests__/inAppMessaging/utils/processInAppMessages.test.ts [hunk 11]
```diff
@@ -53,11 +65,24 @@ describe('processInAppMessages', () => {
...
 			simpleInAppMessagingEvent
 		);
 		expect(result.id).toBe('uuid-3');
 	});
+
+	test('filters in-app messages based on session counts', async () => {
+		// simulate incrementing the counts as this happens when a message is displayed
+		// increment it twice to exceed it's session cap
+		await incrementMessageCounts(messages[2].CampaignId ?? 'uuid-3');
+		await incrementMessageCounts(messages[2].CampaignId ?? 'uuid-3');
```
Comment (cshfang) [inline packages/notifications/__tests__/inAppMessaging/utils/processInAppMessages.test.ts:80]: Nit: Same-ish thought here. I understand that this is done to satisfy strict types but it changes the meaning of the test imo. I think we can either just use `uuid-3` straight up or assert non-null with `!`

File: packages/notifications/jest.config.js [hunk 12]
```diff
@@ -8,7 +8,6 @@ module.exports = {
 	},
 	collectCoverageFrom: ['**/src/**', '!**/src/**/*.native.*'],
 	coverageThreshold: {
-		// TODO(V6): revert these numbers back
```
Comment (cshfang) [inline packages/notifications/jest.config.js:11]: Should we be uncommenting this without taking any action?

File: packages/notifications/src/inAppMessaging/providers/pinpoint/utils/helpers.ts [hunk 13]
```diff
@@ -117,18 +120,20 @@ export const matchesMetrics = (
 	}
 	const memoKey = `${CampaignId}:${JSON.stringify(metrics)}`;
 	if (!eventMetricsMemo.hasOwnProperty(memoKey)) {
-		eventMetricsMemo[memoKey] = Object.entries(Metrics).every(
-			([key, { ComparisonOperator, Value }]) => {
+		eventMetricsMemo[memoKey] =
+			!Metrics ||
```
Comment (cshfang) [inline packages/notifications/src/inAppMessaging/providers/pinpoint/utils/helpers.ts:124]: Nit: Can `??` be used?

File: packages/notifications/src/inAppMessaging/providers/pinpoint/utils/helpers.ts (hunk 7 above)
Comment (cshfang) [reply packages/notifications/src/inAppMessaging/providers/pinpoint/utils/helpers.ts:267]: The Smithy generated types are, afaik, always optional - even if it's not true of the service itself. Without strict mode, we sort of hand-waved this away and said it's actually always there. Imo, providing empty defaults is probably the right thing to do given that the UI was not expected to handle nullish values

File: packages/notifications/src/inAppMessaging/providers/pinpoint/utils/messageProcessingHelpers.ts [hunk 14]
```diff
@@ -107,23 +111,35 @@ async function isBelowCap({
 	);
 
 	return (
-		(!SessionCap || sessionCount < SessionCap) &&
-		(!DailyCap || dailyCount < DailyCap) &&
-		(!TotalCap || totalCount < TotalCap)
+		(!SessionCap || sessionCount <= SessionCap) &&
+		(!DailyCap || dailyCount <= DailyCap) &&
+		(!TotalCap || totalCount <= TotalCap)
```
Comment (cshfang) [inline packages/notifications/src/inAppMessaging/providers/pinpoint/utils/messageProcessingHelpers.ts:116]: Are you sure this is correct (or rather, are you sure this was wrong before?) Iirc, this is supposed to be checking if we're *below* the cap.. If it's *at* the cap, shouldn't the message no longer display? If something changed at the call site of this function - should we rename to `isAtOrBelowCap`?

File: packages/notifications/__tests__/inAppMessaging/providers/pinpoint/apis/setConflictHandler.test.ts (hunk 10 above)
Comment (Samaritan1011001) [reply packages/notifications/__tests__/inAppMessaging/providers/pinpoint/apis/setConflictHandler.test.ts:38]: `messages[0]` will never be returned in this unless the `inAppMessages` change to not have the message with `custom-handled` id anymore.
It's just a fail safe value to return. Don't think it masks any functionality as we are still checking for the customHandledMessage on line 44.

But I changed the index to `messages[2]` which is the custom-handled message in the `inAppMessages`.

File: packages/notifications/__tests__/inAppMessaging/utils/processInAppMessages.test.ts (hunk 11 above)
Comment (Samaritan1011001) [reply packages/notifications/__tests__/inAppMessaging/utils/processInAppMessages.test.ts:80]: used it straight up

Comment (Samaritan1011001) [reply packages/notifications/__tests__/inAppMessaging/utils/processInAppMessages.test.ts]: Non-null asserted it so it fails if the there is no campaignId for messages[2] which is better than assuming the value

File: packages/notifications/__tests__/inAppMessaging/providers/pinpoint/apis/setConflictHandler.test.ts (hunk 10 above)
Comment (Samaritan1011001) [reply packages/notifications/__tests__/inAppMessaging/providers/pinpoint/apis/setConflictHandler.test.ts]: Non-null asserted it so it fails if the there is no matching message which is better than assuming the value

File: packages/notifications/src/inAppMessaging/providers/pinpoint/utils/messageProcessingHelpers.ts (hunk 14 above)
Comment (Samaritan1011001) [reply packages/notifications/src/inAppMessaging/providers/pinpoint/utils/messageProcessingHelpers.ts]: Nothing changed at call site, I assumed this to be the expected behavior, changed it back to just <

**Review by cshfang — APPROVED** [review]

**Review by jimblanc — APPROVED** [review]

**Review by AllanZhengYP — APPROVED** [review]
